File: test_time.kt
import java.util.Calendar

fun main() {
    val times = listOf(
        "02:00", "03:59", "04:01", "08:01", "17:00", "22:00"
    )
    for (time in times) {
        val currentCal = Calendar.getInstance()
        val parts = time.split(":")
        currentCal.set(Calendar.HOUR_OF_DAY, parts[0].toInt())
        currentCal.set(Calendar.MINUTE, parts[1].toInt())
        currentCal.set(Calendar.SECOND, 0)
        
        val startCal = currentCal.clone() as Calendar
        startCal.set(Calendar.HOUR_OF_DAY, 21)
        startCal.set(Calendar.MINUTE, 0)
        
        val endCal = currentCal.clone() as Calendar
        endCal.set(Calendar.HOUR_OF_DAY, 4)
        endCal.set(Calendar.MINUTE, 0)
        
        if (endCal.before(startCal) || endCal == startCal) {
            val cutoffCal = endCal.clone() as Calendar
            cutoffCal.add(Calendar.HOUR_OF_DAY, 4)
            if (currentCal.before(cutoffCal)) {
                startCal.add(Calendar.DAY_OF_YEAR, -1)
            } else {
                endCal.add(Calendar.DAY_OF_YEAR, 1)
            }
        }
        
        val isCurrent = !currentCal.before(startCal) && currentCal.before(endCal)
        val isPast = !currentCal.before(endCal)
        
        println("Time $time -> Past: $isPast, Current: $isCurrent")
    }
}

File: test_time.py
import datetime

times = ["02:00", "03:59", "04:01", "08:01", "17:00", "22:00"]
today = datetime.datetime.now().replace(second=0, microsecond=0)

for t in times:
    hr, mn = map(int, t.split(':'))
    current = today.replace(hour=hr, minute=mn)
    
    start = current.replace(hour=21, minute=0)
    end = current.replace(hour=4, minute=0)
    
    if end <= start:
        cutoff = end + datetime.timedelta(hours=4)
        if current < cutoff:
            start -= datetime.timedelta(days=1)
        else:
            end += datetime.timedelta(days=1)
            
    isCurrent = (current >= start) and (current < end)
    isPast = (current >= end)
    print(f"Time {t} -> Past: {isPast}, Current: {isCurrent}")
